flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarOrderedSourceReaderTest.java

```diff
@@ -0,0 +1,167 @@
+                (PulsarOrderedSourceReader<Integer>) sourceReader(autoAcknowledgementEnabled);
+        String topicName = uniqueTestName(testInfo);
+        setupSourceReader(reader, topicName, 0, Boundedness.CONTINUOUS_UNBOUNDED);
+
+        // waiting for results
+        TestingReaderOutput<Integer> output = new TestingReaderOutput<>();
+        pollUntil(
+                reader,
+                output,
+                () -> output.getEmittedRecords().size() == NUM_RECORDS_PER_PARTITION,
+                "The output didn't poll enough records before timeout.");
+        reader.snapshotState(100L);
+        reader.notifyCheckpointComplete(100L);
+        pollUntil(
+                reader,
+                output,
+                reader.cursorsToCommit::isEmpty,
+                "The offset commit did not finish before timeout.");
+
+        // verify consumption
+        reader.close();
+        verifyAllMessageAcknowledged(
+                NUM_RECORDS_PER_PARTITION, TopicNameUtils.topicNameWithPartition(topicName, 0));
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void offsetCommitOnCheckpointComplete(boolean autoAcknowledgementEnabled, TestInfo testInfo)
+            throws Exception {
+        PulsarOrderedSourceReader<Integer> reader =
+                (PulsarOrderedSourceReader<Integer>) sourceReader(autoAcknowledgementEnabled);
+        String topicName = uniqueTestName(testInfo);
+        // consume more than 1 partition
+        reader.addSplits(
+                createPartitionSplits(
+                        topicName, DEFAULT_PARTITIONS, Boundedness.CONTINUOUS_UNBOUNDED));
+        reader.notifyNoMoreSplits();
+        TestingReaderOutput<Integer> output = new TestingReaderOutput<>();
+        long checkpointId = 0;
+        int emptyResultTime = 0;
+        InputStatus status;
+        do {
+            checkpointId++;
+            status = reader.pollNext(output);
+            // Create a checkpoint for each message consumption, but not complete them.
+            reader.snapshotState(checkpointId);
+            // the first couple of pollNext() might return NOTHING_AVAILABLE before data appears
+            if (InputStatus.NOTHING_AVAILABLE == status) {
+                emptyResultTime++;
+                sleepUninterruptibly(1, TimeUnit.SECONDS);
+            }
+
+        } while (emptyResultTime < MAX_EMPTY_POLLING_TIMES
+                && status != InputStatus.END_OF_INPUT
+                && output.getEmittedRecords().size()
+                        < NUM_RECORDS_PER_PARTITION * DEFAULT_PARTITIONS);
+
+        // The completion of the last checkpoint should subsume all previous checkpoints.
+        assertEquals(checkpointId, reader.cursorsToCommit.size());
+
+        long lastCheckpointId = checkpointId;
+        CommonTestUtils.waitUtil(
+                () -> {
+                    try {
+                        reader.notifyCheckpointComplete(lastCheckpointId);
```

Comment:
I think the logic is when the checkpoint complete and the  cursorsToCommit is empty, then it means the checkpoint is OK. It will run multiple times if there is an exception. Ideally after one run the cursorsToCommit should become empty. But I do agree maybe we can optimize this so that if one exception is thrown, we consider the test as fail.